### ISSUE TITLE
Improve report-aggregate documentation and behaviour

### ISSUE THREAD

**Issue by skjolber** [jacoco/jacoco#974]
### Scenario

* JaCoCo version: 0.8.5 (latest)
* Operating system: Linux
* Tool integration: Maven
* Description of your use case: create aggregate report for multi-module Maven build

### Current Behaviour
First report-aggregate creates an empty report. Then the documentation page at [report-aggregate-mojo](https://www.jacoco.org/jacoco/trunk/doc/report-aggregate-mojo.html) is without examples, and does not link to your integration test examples. When finally getting a working example, adjusting it to get simple things like `dataFileIncludes` working is generally hard or (for me) impossible. This, and issue #812.

A solution where creating a new Maven project for generating a report will come as a surprise on 99% of developers. So the task interface is frankly broken IMHO.

### Wanted Behaviour
A task which works with zero configuration, and where modules are included when packaging is pom, dependencies when packaging is jar.

### Possible Workarounds
An aggregate task which runs in a second invocation of Maven. Create a new task if legacy is a problem, i.e. `report-aggregate-files` or something like that.

**Comment by marchof** [comment]
I agree that this goal has no nice "out of the box" experience". But this is mainly due to the maven model which has not proper support for aggregate scenarios.

> where modules are included when packaging is pom

Unfortunately this does not work because the parent project is executed first. You can read about our implementation attempts here: https://github.com/jacoco/jacoco/wiki/MavenMultiModule

**Comment by skjolber** [comment]
Nice documentation. But if the parent project must run to its completion (all phases / tasks) first, then why not simply support generation of the aggregate report using a second maven invokation (when all files .exec files are on disc)? Or am I missing something here?

**Comment by marchof** [comment]
Not sure if it is a better "out of the box experience" when we tell users that they need to trigger their builds twice with different parameters.

**Comment by skjolber** [comment]
In a nutshell, when I've invested time and effort in creating my project 'the right way' by using modules, it feels so 'the wrong way' to add a side module just to generate a report. The report module will be visible in the source tree, has to be documented and so on. It'll be anoying, but I'll rather run maven twice.

But it would not necessarily require the build to run twice, would it, if not tied to the lifecycle? 

Would creating a new experimental task be interesting, and then just letting users decide which approach they prefer? This would strictly generate from files already present in modules. Ideally working for something like

> mvn jacoco:report-files-aggregate -N

**Comment by bmaehr** [comment]
As far as I see Sonar has the same problem and seems to have solved it somehow. I'm not sure which of the configuration here dos ist, but it works: https://github.com/SonarSource/sonar-scanner-maven/blob/master/src/main/java/org/sonarsource/scanner/maven/SonarQubeMojo.java

**Comment by bmaehr** [comment]
Additionally this problem gets more urgent, because Sonarqube 8.x is not supporting the jacoco native format anymore, so the generation of a aggregated xml file for sonar is necessary.

**Comment by peternelissen** [comment]
The latest SonarQube version 8.2 does not work with the exec files anymore.

I also agree that we should not be forced to add a new module just for code coverage.

The report-aggregate goal should include indirect maven dependencies (and not only direct maven module dependencies). Ideally also itself, but ok, adding normal report goal solves that.

Then the report-aggregate can just be added to the maven module that builds your final war/jar, and it will be ok.

I've created following sonarsource issue for this: https://community.sonarsource.com/t/jacoco-xml-format-cannot-work-as-good-as-the-binary-format-for-multi-module-maven/21293/8

It has an example why the XML format does not work: https://github.com/peternelissen/jacoco-sonar

**Comment by peternelissen** [comment]
@marchof Should I open a new issue for my comment? (fact that report-aggregate does not include indirect maven dependencies) This issue would also solve it, if all modules would be added by default if we can just add it to the master pom.

**Comment by bmaehr** [comment]
No, it is not necessary to add a new issue for a workaround for the broken design, as I already implemented a working clean solution (see comment of 5th Mar)

**Comment by dan323** [comment]
Just as an example, [pitest](https://github.com/hcoles/pitest) created a new a *report-aggregate-module* goal to do this aggregation through pom dependencies executed at the end of the lifecycle, while keeping the *report-aggregate* goal similar to the one in jacoco.

**Comment by bmaehr** [comment]
> Just as an example, [pitest](https://github.com/hcoles/pitest) created a new a _report-aggregate-module_ goal to do this aggregation through pom dependencies executed at the end of the lifecycle, while keeping the _report-aggregate_ goal similar to the one in jacoco.

I already implemented it for jacoco, see above.

**Comment by dan323** [comment]
> > Just as an example, [pitest](https://github.com/hcoles/pitest) created a new a _report-aggregate-module_ goal to do this aggregation through pom dependencies executed at the end of the lifecycle, while keeping the _report-aggregate_ goal similar to the one in jacoco.
> 
> I already implemented it for jacoco, see above.

Is it merge as a new goal? I saw you provided a the implementation in a comment, but not that it is going to be released. My example is on how to provide this feature without impacting who already is using the aggregate goal.

**Comment by bmaehr** [comment]
> > > Just as an example, [pitest](https://github.com/hcoles/pitest) created a new a _report-aggregate-module_ goal to do this aggregation through pom dependencies executed at the end of the lifecycle, while keeping the _report-aggregate_ goal similar to the one in jacoco.
> > 
> > 
> > I already implemented it for jacoco, see above.
> 
> Is it merge as a new goal? I saw you provided a the implementation in a comment, but not that it is going to be released. 

No, it is not MR and it is no new goal. But I think it is a todo of 5 minutes to attach it to a new goal, if the owners of the project decide to do that after a long discussion. At least much less effort then searching something in some other project how it works.

> My example is on how to provide this feature without impacting who already is using the aggregate goal.

You didn't provide an example or implementation, you provided some reference to an other project.

**Comment by dan323** [comment]
> > > > Just as an example, [pitest](https://github.com/hcoles/pitest) created a new a _report-aggregate-module_ goal to do this aggregation through pom dependencies executed at the end of the lifecycle, while keeping the _report-aggregate_ goal similar to the one in jacoco.
> > > 
> > > 
> > > I already implemented it for jacoco, see above.
> > 
> > 
> > Is it merge as a new goal? I saw you provided a the implementation in a comment, but not that it is going to be released.
> 
> No, it is not MR and it is no new goal. But I think it is a todo of 5 minutes to attach it to a new goal, if the owners of the project decide to do that after a long discussion. At least much less effort then searching something in some other project how it perhaps works.
> 
> > My example is on how to provide this feature without impacting who already is using the aggregate goal.
> 
> You didn't provide an example or implementation, you provided some reference to an other project.

It was just an example of some plugin that had the same issue and made a new goal just adding "-module" to the legacy goal name. My comment was not intended to give an implementation, I had already seen yours and not necessary to go further on that as it works.

**Comment by bmaehr** [comment]
Ok, then I understood your intention wrong. Thank you for clarification. 

Let the discussion about creating a new goal or modifying the existing one begin :-)
If the question is decided I can provide a MR.

**Comment by peternelissen** [comment]
As the behaviour is very different from the existing goal, there is no way to be backwards compatible. 

A new goal sounds perfect. report-aggregate-module, or report-all

Who can decide on this? 

@marchof I assume you can help us our if it's realistic if this can continue? Thx!

**Comment by marchof** [comment]
Dear @peternelissen, to be honest it we will probably not add new features to our Maven plug-in. We simply cannot afford to support this. Check our mailing list, there is so many questions about how Maven works.

Ideally a separate project picks up the JaCoCo Maven integration and we can focus on the coverage engine ... just dreaming 😃

**Comment by skjolber** [comment]
Perhaps linking to the mailing list via the docs would be better, at least for the main readme.. 

And a note on using jacoco with sonar, [this is working](https://github.com/entur/jwt-resource-server/blob/master/.travis.yml#L21) for me by using `find` and a command-line properties.

**Comment by nightswimmings** [comment]
Please, this is the most important hole of an enterprise-grade pom right now. We dedicated too much time to it.
As soon as we all migrate to new Sonar, we all get lower coverage and are forced to add tests on shared modules, even if we add the report-aggregate goal, simply because the jacoco-aggregate report is not merged with the others in an OR fashion. We are completely stubborn on not starting adding and mantaining new modules to our multimodule projects simply because this pitfall

**Comment by marchof** [comment]
@nightswimmings You get lower coverage because what you currently see is just "accidental coverage" (but no explicit tests). What about encouraging your team to actually write unit tests for the new modules? 😉

**Comment by nightswimmings** [comment]
When we have a module where we place all the abstractions that all other modules use, and these client modules have a 100% Integration Tests coverage it is an overkill to test the abstract classes in the common module, since this is tested as many as 5 times through the others

**Comment by nightswimmings** [comment]
But to be honest, I think it's Sonar the one that should be able to interpret a logical OR when specifying multiple reports on the xmlReportPaths var (including the IT, the UT, and the aggregate for coverage over common modules), even if a report-all goal it is not implemented in Jacoco

**Comment by bmaehr** [comment]
I uploaded a patched version of the jacoco plugin with a fixed (=works like users would expect) report-aggregate goal. Just rename jacoco-maven-plugin-0.8.5-PATCH.zip to jacoco-maven-plugin-0.8.5-PATCH.jar and add it to your local maven repository.
[jacoco-maven-plugin-0.8.5-PATCH.zip](https://github.com/jacoco/jacoco/files/4662378/jacoco-maven-plugin-0.8.5-PATCH.zip)

**Comment by tpischke-bedag** [comment]
I'd like to add my voice to the chorus contending that forcing projects to introduce a new reporting module is not an acceptable solution. We need a solution that doesn't lose the existing functionality we had with the binary format. That is to say, we need a solution that works out of the box, without the creation of new maven projects or the maintenance of an unwieldy property listing all the XML report locations. Otherwise, it will be difficult to convince our developers to make this effort in every new project 'just' for the jacoco tool.

I understand that the XML Format is more convenient for you, but placing additional burden on users without adding any additional value is going to be a hard sell. Dropping support for the binary format also places projects in a very uncomfortable position: Invest time and money and add complexity just  to maintain the status quo, or drop the functionality until a better alternative is available.

**Comment by tpischke-bedag** [comment]
@bmaehr Will the patch referenced in https://github.com/jacoco/jacoco/issues/974#issuecomment-632055670 be officially released soon?

**Comment by alban-auzeill** [comment]
Hello JaCoCo team, SonarQube java analyzer team speaking,

For sure, we don't want to depend anymore from JaCoCo's native format (.exec), that's why, since SonarQube 8.2, we only depend on the XML format.

Unfortunately, we discovered that some of our users are not able to use the XML format, because they fail to aggregate some coverage reports, like @peternelissen above.

Disclaimer: We believe that aggregating coverage of several maven modules blurs the coverage report by adding some accidental coverage without assertions. So we think it's a bad practice and we would not suggest it.

Still, because it was possible to append several coverage results in the same JaCoCo native file, some users did it. They merged the coverage of several Maven modules and now they are forced to do the same using JaCoCo XML format to keep the same level of coverage.

Unfortunately, the "report-aggregate" of "jacoco-maven-plugin" only aggregates reports of direct dependencies and not transitive ones. Because of this limitation, it can not be used in all cases.

At SonarSource, we tried to find a solution for users like @peternelissen using JaCoCo 0.8.5, but we failed. We are always stuck at the same point. We have the coverage information that we want in the native format, but this information is filtered when we produce the XML report. Whatever the change we could do on SonarQube side, it will not be possible to merge data that we don’t have.

So, we need your help to choose what we should do:
- We should contribute to JaCoCo to add an option to "report-aggregate" to include transitive dependent modules.
- Force our users to change their direct dependencies so that "report-aggregate" includes the coverage of those dependencies.
- Tell our user that there is no solution, "report-aggregate" has some drawbacks, and should be used with care.

Cheers,

Alban and @quentin-jaquier-sonarsource

Related discussions at SonarSource:
- https://community.sonarsource.com/t/21293
- https://community.sonarsource.com/t/26065

**Comment by bmaehr** [comment]
@alban-auzeill Could you describe what's the problem with dependencies?

**Comment by alban-auzeill** [comment]
@bmaehr, there's a small code example made by @peternelissen that reproduce the dependency problem: [jacoco-example4](https://github.com/peternelissen/jacoco-sonar/tree/master/jacoco-example4)
We failed to see line 14 of `M1.java` as covered. (It is called by M3Test)
There are 3 maven modules: `jacoco-m1`, `jacoco-m2`, `jacoco-m3`
Dependencies are `jacoco-m3 -> jacoco-m2 -> jacoco-m1`
But there's no explicit dependency `jacoco-m3 -> jacoco-m1`, it is only a transitive one. 
Execute `mvn clean install` and look at `jacoco-m3/target/site/jacoco-aggregate/jacoco.xml`. There's no coverage information about `M1.java`.

Making transitive dependencies explicit fixes the problem. Uncomment the `jacoco-m1` dependency in `jacoco-m3/pom.xml`, execute `mvn clean install` and look at `jacoco-m3/target/site/jacoco-aggregate/jacoco.xml`. You will see line 14 of `M1.java` as covered.

@peternelissen's need is to adapt JaCoCo configuration to produce XML aggregated report without changing module dependencies. And one solution is to change the way ReportAggregateMojo collect dependencies to aggregate.

**Comment by bmaehr** [comment]
@alban-auzeill I think this has nothing to do with this issue.

**Comment by peternelissen** [comment]
@alban-auzeill Thanks for looking into this!

An extra comment: with the native format, all modules could just write to the same .exec file in the parent folder (there was a parameter for this, to add to the file). It was not needed to merge seperate modules. This made the native format very easy to use.

Changing the report-aggregate to include transitive dependencies would solve the problem partially. Users would still need to add another goal for that module, because report-aggregate doesnt include coverage of itself. The config will get error prone for complicated maven structures. (projects with two WAR files for example, that both depend on the same sub modules)

The proposed change of @bmaehr is better and very user friendly: a goal that runs at the end of the maven lifecycle and covers ALL modules of that project. That's exactly what users want. You just have the coverage of **all code** and **all tests** of **all maven modules** of that project!

Report-aggregate works very different, it's not possible to do this change backwards compatible. A new goal is needed.

**Comment by alban-auzeill** [comment]
@bmaehr I agree that the problem I wanted to fix is not exactly the one described in the first comment of this issue.
It only related because of "Improve report-aggregate ... behaviour", "create aggregate report for multi-module Maven build", "A solution where creating a new Maven project for generating a report will come as a surprise on 99% of developers."

But then, some comments diverged to focus on aggregating all dependent modules. I tested your March 5th `ReportAggregateMojo.java.txt` modification, and it accidentally fixes the above [jacoco-example4](https://github.com/peternelissen/jacoco-sonar/tree/master/jacoco-example4) coverage problem.

When @peternelissen asked "Should I open a new issue for my comment?", your answer was "No, ... I already implemented a working clean solution"
  
So when I see your comment: "I think this has nothing to do with this issue.", I'm lost. Should I create another issue or forget about this problem, I have no clue.

**Comment by alban-auzeill** [comment]
@peternelissen For now, I'm just trying to clarify the problem and ask JaCoCo team if they wish to have it solved.
I have not thought if changing `report-aggregate` is the best solution.

**Comment by arey** [comment]
@alban-auzeill I prefer the first option: 

> We should contribute to JaCoCo to add an option to "report-aggregate" to include transitive dependent modules

It may be an optional parameter with the transitive dependencies scan disabled.

**Comment by andpab** [comment]
The dependency problem that is mentioned in this issue has already been described in more detail (along with a possible solution) in #1050

**Comment by nightswimmings** [comment]
@alban-auzeill
``` Disclaimer: We believe that aggregating coverage of several maven modules blurs the coverage report by adding some accidental coverage without assertions. So we think it's a bad practice and we would not suggest it.```

Just to double-check. Do you think, from Sonar, that all abstract code on commons module which is covered through the implementations on dependent modules should have coverage on its own?

**Comment by arey** [comment]
To complete the @nightswimmings question: same things with getter/setter of DTO, enums, exceptions constructor and all common stuff.

**Comment by peternelissen** [comment]
Should jacoco take an opiniated view on this?

I think jacoco should allow both. And the industry can decide what's best. It can also depend on your module split, your design and your testing strategy. 

For example, I know teams that have two reports: one for unit tests and one for integration tests (spring boot tests / tests using REST API for example). For unit tests I agree that you dont want coverage across modules. But for integration tests?

I think it's dangerous to make assumptions and limit teams to the test strategy defined by this community.

**Comment by eeverman** [comment]
Some examples of best practices on the [MavenMultiModule](https://github.com/jacoco/jacoco/wiki/MavenMultiModule) page and featured other places in the docs would be really helpful.

I get it - Maven doesn't have a good solution for this.  But consider the path of the average developer:
1.  Has a multi-module project and wants coverage reports (this is everyone)
2. Not super familiar w/ Jacoco, but grab the plugin (basically everyone)
3. Gosh, the setup looks complicated... Why do I need multiple executions? (basically everyone)
4. OK, I think I get it.  But do I choose `report` or `report-aggregate`??  Hmm... oh, I see:
> report-aggregate:  Creates a structured code coverage report (HTML, XML, and CSV) from multiple projects within reactor.

Great!  I'm all set for my multi-module project!

...and it doesn't quite seem to be working right.  My reports to CodeCov are missing some classes, coverage numbers are off, etc..

Some prominent warnings / notices of **"Hey!  This well marked expressway marked 'Multi-Projects this way' is not the correct path."**  Look at this example.  Understand the difference between a report and an aggregate.  See this discussion.

**Comment by bmaehr** [comment]
Patched version of 0.8.
[jacoco-maven-plugin-0.8.7-PATCH.jar.zip](https://github.com/jacoco/jacoco/files/7466627/jacoco-maven-plugin-0.8.7-PATCH.jar.zip)
7

**Comment by reda-alaoui** [comment]
@bmaehr Could you provide the sources of your jar? Or even better, could you fork and create a pull request?

**Comment by bmaehr** [comment]
> @bmaehr Could you provide the sources of your jar? Or even better, could you fork and create a pull request?

It is not so easy, because it changes the behaviour of an existing target.

**Comment by ambar-cap** [comment]
Sorry if this question does not make sense on this thread. I came across this while trying to figure out why my coverage report does not have the correct numbers in a multi module maven setup.

Can someone please tell the exact steps and dependency versions needed for the multi module maven setup to work correctly? I have tried to follow the following guide -- [[Coverage & Test Data] Importing JaCoCo coverage report in XML format](https://community.sonarsource.com/t/coverage-test-data-importing-jacoco-coverage-report-in-xml-format/12151) to setup my project, but the coverage numbers seem to be off.

Also, I am unable to make jacoco ignore the files from coverage and no matter what i try, the coverage numbers don't seem to budge. I have tried adding jacoco exclusions in every module, tried adding sonar exclusion property in poms of all modules, and also tried adding exclusions via the sonar UI but nothing seems to be working. Although the `report` goal for individual projects lists the reduced number of files in the following log line -- `Analyzed bundle 'MyTestProject' with 45 classes`, but at the end of `report-aggregate` goal, this numbers seems to be getting reset again to the original number of files (without any exclusions). Can someone please tell me what might be the correct way to do this.

Thanks in advance and any help would be appreciated :)

cc: @bmaehr @alban-auzeill

**Comment by m-balakhontsev** [comment]
Current design seems really inconvenient. Hope it will be eventually changed as proposed in this issue.

**Comment by CharlieReitzel** [comment]
> Current design seems really inconvenient. Hope it will be eventually changed as proposed in this issue.

Agreed, the current design is a constant wrestling match to get it working.  And I keep losing ... :--(

**Comment by tircis** [comment]
Hi,
Since I was struggling with same problem (Jacoco plugin doesn't following maven's reactor to collect its dependencies, in particular with multi-modules project), and reading a bunch of articles, i decided to fix it on my own with a dedicated maven-plugin that mimics Jacoco official one. Please have a look at it and give me some feedback about it : [jacoco-aggregate-submodule](https://github.com/codefilarete/jacoco-aggregate-submodule)

**Comment by velo** [comment]
> Ideally a separate project picks up the JaCoCo Maven integration and we can focus on the coverage engine ... just dreaming 😃

Well, hehhe, I just did that
https://github.com/velo/easy-jacoco-maven-plugin

My extension just turn enable jacoco on maven projects with zero configuration and produce reports/checks project wide.

**Comment by shencangsheng** [comment]
@velo An exceptionally useful plugin.
